File: src/main/java/couponsProjectPhase3/services/AdminService.java
package couponsProjectPhase3.services;

import couponsProjectPhase3.beans.Admin;
import couponsProjectPhase3.beans.Category;
import couponsProjectPhase3.beans.Company;
import couponsProjectPhase3.beans.Customer;
import couponsProjectPhase3.exceptions.*;
import couponsProjectPhase3.exceptions.unallowedUpdateExceptions.*;
import couponsProjectPhase3.repositories.*;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AdminService extends ClientService {
    private int adminId;
    private AdminRepository adminRepository;

    //ctor
    public AdminService(CompaniesRepository companiesRepository, CouponsRepository couponsRepository, CustomersRepository customersRepository, CategoriesRepository categoriesRepository, AdminRepository adminRepository) {
        super(companiesRepository, couponsRepository, customersRepository, categoriesRepository);
        this.adminRepository = adminRepository;
    }

    //methods
    public boolean login(String email, String password) throws EmptyValueException {
        Admin admin = adminRepository.findByEmailAndPassword(email, password).orElse(null);
        if (admin != null) {
            adminId = admin.getId();
            return true;
        }
        return false;
    }

    public Company addCompany(Company company) throws  EmptyValueException, UnallowedUpdateException {

        //we'll check whether we've got an empty value which isn't allowed to be empty
        if (company == null || company.getName() == null || company.getName().isEmpty() || company.getPassword() == null
                || company.getPassword().isEmpty() || company.getEmail() == null || company.getEmail().isEmpty())
            throw new EmptyValueException();

        //we are not allowed to insert a new company with an id other than 0 - the user cannot choose the company's id
        //because it is auto-generated
        if (company.getId() != 0)
            throw new UnallowedUpdateException();

        //we'll check whether the input email and password match our format properly
        if (!Validations.isValidEmail(company.getEmail()))
            throw new EmailFormatException();

        if (!Validations.isValidPassword(company.getPassword()))
            throw new PasswordFormatException();

        //we'll check whether there's already a company in the db with the same name or email
        if (companiesRepository.existsByEmail(company.getEmail()) || companiesRepository.existsByName(company.getName()))
            throw new AlreadyExistingValueException();

        return companiesRepository.save(company);
    }

    public Company updateCompany(Company company) throws EmptyValueException,
            UnallowedUpdateException, NonexistantObjectException, AlreadyExistingValueException, EmailFormatException, PasswordFormatException {
        if (company == null)
            throw new EmptyValueException();

        //we cannot update a nonexistant company - we'll check if this company actually exists
        Company companyInDB = companiesRepository.findById(company.getId()).orElse(null);

        if (companyInDB == null)
            throw new NonexistantObjectException();

        //we'll check whether we've got an empty value which isn't allowed to be empty
        if (company.getName() == null || company.getName().isEmpty() || company.getPassword() == null
                || company.getPassword().isEmpty() || company.getEmail() == null || company.getEmail().isEmpty())
            throw new EmptyValueException();

        //we'll check whether the input email and password match our format properly
        if (!Validations.isValidEmail(company.getEmail()))
            throw new EmailFormatException();

        if (!Validations.isValidPassword(company.getPassword()))
            throw new PasswordFormatException();

        //we cannot update to an email that already exists in the db -
        // we'll check first if the user is actually trying to change the company's email and then well look
        // for matching values
        if (!company.getEmail().equals(companyInDB.getEmail()) &&
                companiesRepository.existsByEmail(company.getEmail()))
            throw new AlreadyExistingValueException();

        //we are not allowed to update name
        if (!companyInDB.getName().equals(company.getName()))
            throw new UnallowedUpdateException();

        //we cannot update a company while adding it new coupons - we'll reset its coupons list
        company.setCoupons(null);

        return companiesRepository.save(company);

    }

    public void deleteCompany(int companyId) throws NonexistantObjectException {
        //we cannot delete a nonexistant company
        if (!companiesRepository.existsById(companyId))
            throw new NonexistantObjectException();
        //we have to delete also the coupons issued by the company and its customer purchase history
        companiesRepository.deleteCouponsPurchaseHistory(companyId);
        couponsRepository.deleteAllCouponsByCompanyId(companyId);
        companiesRepository.deleteById(companyId);


    }

    public List<Company> getAllCompanies() {
        return companiesRepository.findAll();
    }

    public Company getOneCompany(int companyId) {
        return companiesRepository.findById(companyId).orElseThrow();
    }

    public Customer addCustomer(Customer customer) throws EmptyValueException, UnallowedUpdateException {
        //we'll check for disallowed empty values
        if (customer == null || customer.getEmail() == null || customer.getEmail().isEmpty() ||
                customer.getPassword() == null || customer.getPassword().isEmpty() || customer.getFirstName() == null ||
                customer.getFirstName().isEmpty() || customer.getLastName() == null || customer.getLastName().isEmpty())
            throw new EmptyValueException();

        //we can't add a new customer that already has an id. it should be auto-generated
        if (customer.getId() != 0)
            throw new UnallowedUpdateException();

        //we'll verify the password and email format
        if (!Validations.isValidPassword(customer.getPassword()))
            throw new PasswordFormatException();
        if (!Validations.isValidEmail(customer.getEmail()))
            throw new EmailFormatException();

        //we'll check whether the customer's name is valid
        if (!Validations.onlyAlphabets(customer.getFirstName()) || !Validations.onlyAlphabets(customer.getLastName()))
            throw new NameException();


        //we mustn't add a new customer with an existing email address
        if (customersRepository.existsByEmail(customer.getEmail()))
            throw new AlreadyExistingValueException();

        return customersRepository.save(customer);
    }

    public Customer updateCustomer(Customer customer) throws EmptyValueException, NonexistantObjectException,
            AlreadyExistingValueException, PasswordFormatException, EmailFormatException, NameException {
        //we cannot accept a null expression
        if (customer == null)
            throw new EmptyValueException();

        //we cant update a nonexistant customer
        if (!customersRepository.existsById(customer.getId()))
            throw new NonexistantObjectException();

        //we'll check for disallowed empty values
        if (customer.getEmail() == null || customer.getEmail().isEmpty() ||
                customer.getPassword() == null || customer.getPassword().isEmpty() || customer.getFirstName() == null ||
                customer.getFirstName().isEmpty() || customer.getLastName() == null || customer.getLastName().isEmpty())
            throw new EmptyValueException();

        //we cannot update a customer email to an already existing email -
        // we'll check first if the user is actually trying to change the customer's email and then well look
        // for matching values
        if (!customer.getEmail().equals(customersRepository.findById(customer.getId()).get().getEmail()) &&
                customersRepository.existsByEmail(customer.getEmail()))
            throw new AlreadyExistingValueException();

        //we'll verify the password and email format
        if (!Validations.isValidPassword(customer.getPassword()))
            throw new PasswordFormatException();
        if (!Validations.isValidEmail(customer.getEmail()))
            throw new EmailFormatException();

        //we'll check whether the customer's name is valid
        if (!Validations.onlyAlphabets(customer.getFirstName()) || !Validations.onlyAlphabets(customer.getLastName()))
            throw new NameException();

        //we cannot update customer id
       return customersRepository.save(customer);
    }

    public void deleteCustomer(int customerId) throws NonexistantObjectException {
        //we cannot delete a nonexistant customer
        if (!customersRepository.existsById(customerId))
            throw new NonexistantObjectException();

        //we have to delete their coupons purchase history as well
        customersRepository.deleteCouponsPurchaseHistory(customerId);
        customersRepository.deleteById(customerId);
    }

    public List<Customer> getAllCustomers() {
        return customersRepository.findAll();
    }

    public Customer getOneCustomer(int customerId) {
        return customersRepository.findById(customerId).orElseThrow();
    }

    public List<Category> getCategories() {
        return categoriesRepository.findAll();
    }

    public Category getOneCategory(int categoryId) {
        return categoriesRepository.findById(categoryId).orElseThrow();
    }

    public Category addCategory(Category category) throws UnallowedUpdateException, EmptyValueException {
        //we cannot add a null value
        if (category == null)
            throw new EmptyValueException();

        //we cannot add a category that is not only alphabetical
        if (!Validations.onlyAlphabets(category.getName()))
            throw new NameException();

        //we cannot add category with an already existing name
        if(categoriesRepository.existsByName(category.getName()))
            throw new AlreadyExistingValueException();

        //we cannot add a category with a self-chosen id
        if (category.getId() != 0)
            throw new UnallowedUpdateException();

        return categoriesRepository.save(category);
    }

    public Category updateCategory(Category category) throws EmptyValueException, NonexistantObjectException, NameException, AlreadyExistingValueException {
        //we cannot update a null value
        if (category == null)
            throw new EmptyValueException();

        //we cannot update a nonexistant category
        if (!categoriesRepository.existsById(category.getId()))
            throw new NonexistantObjectException();

        //we cannot update a category whose name is not only alphabetical
        if (!Validations.onlyAlphabets(category.getName()))
            throw new NameException();

        //we cannot update category with an already existing name
        if(categoriesRepository.existsByName(category.getName()))
            throw new AlreadyExistingValueException();


        return categoriesRepository.save(category);
    }

    //right now I do not want to implement a deleteCategory method because it is too risky. Categories can be deleted
    // directly through the DB

    public Admin getAdminDetails() {
        return adminRepository.findById(adminId).orElse(null);
    }
}

File: src/services/AdminService.ts
import axios from "axios";
import { store } from "../store";
import { Company } from "../models/Company";
import { addCompany, deleteCompany, fetchCompanies, updateCompany } from "../reducers/CompaniesSlice";
import { Customer } from "../models/Customer";
import { Category } from "../models/Category";
import { addCategory, fetchCategories, updateCategory } from "../reducers/CategoriesSlice";
import { Admin } from "../models/Admin";

class AdminService {

    async getAdminDetails() {
        try {
            return (await axios.get<Admin>("http://localhost:8080/admin")).data;
        } catch (error) {
            console.error("Error fetching admin:", error);
            throw error; // Propagate the error to be handled in the calling component
        }
    }

    async addCompany(company: Company) {
        try {
            const response = await axios.post<Company>("http://localhost:8080/admin/company", company);
            store.dispatch(addCompany(response.data)); // Update Redux state with the added company
            return response.data; // Return the newly added company
        } catch (error) {
            console.error("Error adding company:", error);
            throw error; // Propagate the error to be handled in the calling component
        }
    }

    async updateCompany(company: Company) {
        try {
            const response = await axios.put<Company>("http://localhost:8080/admin/company", company);
            store.dispatch(updateCompany(response.data)); // Update Redux state with the updated company
            return response.data; // Return the updated company
        } catch (error) {
            console.error("Error updating company:", error);
            throw error;
        }
    }

    async deleteCompany(id: number) {
        try {
            const response = await axios.delete<Company>(`http://localhost:8080/admin/company/${id}`);
            store.dispatch(deleteCompany(id)); // Remove company from Redux state
            return response.data; // Return deleted company
        } catch (error) {
            console.error("Error deleting company:", error);
            throw error;
        }
    }

    async getAllCompanies() {
        try {
            // Check if companies are already in Redux store
            const companies = store.getState().companies.companies;
            if (companies.length === 0) {
                const response = await axios.get<Company[]>("http://localhost:8080/admin/companies");
                store.dispatch(fetchCompanies(response.data)); // Fetch and update Redux store with companies
                return response.data; // Return the fetched companies
            }
            return companies; // Return companies from Redux store if already fetched
        } catch (error) {
            console.error("Error fetching companies:", error);
            throw error;
        }
    }

    async getOneCompany(id: number) {
        try {
            let company = store.getState().companies.companies.find(c => c.id === id);
            if (!company) {
                company = (await axios.get<Company>(`http://localhost:8080/admin/company/${id}`)).data;
            }
            return company; // Return found or fetched company
        } catch (error) {
            console.error("Error fetching company:", error);
            throw error;
        }
    }

    // Customer-related methods
    async addCustomer(customer: Customer) {
        try {
            const response = await axios.post("http://localhost:8080/admin/customer", customer);
            return response.data; // Return added customer
        } catch (error) {
            console.error("Error adding customer:", error);
            throw error;
        }
    }

    async updateCustomer(customer: Customer) {
        try {
            const response = await axios.put("http://localhost:8080/admin/customer", customer);
            return response.data; // Return updated customer
        } catch (error) {
            console.error("Error updating customer:", error);
            throw error;
        }
    }

    async deleteCustomer(id: number) {
        try {
            const response = await axios.delete<Customer>(`http://localhost:8080/admin/customer/${id}`);
            return response.data; // Return deleted customer
        } catch (error) {
            console.error("Error deleting customer:", error);
            throw error;
        }
    }

    async getAllCustomers() {
        try {
            const response = await axios.get<Customer[]>('http://localhost:8080/admin/customers');
            return response.data; // Return all customers
        } catch (error) {
            console.error("Error fetching customers:", error);
            throw error;
        }
    }

    async getOneCustomer(id: number) {
        try {
            const response = await axios.get<Customer>(`http://localhost:8080/admin/customer/${id}`);
            return response.data; // Return one customer
        } catch (error) {
            console.error("Error fetching customer:", error);
            throw error;
        }
    }

    // Category-related methods
    async addCategory(category: Category) {
        try {
            const response = await axios.post<Category>("http://localhost:8080/admin/category", category);
            store.dispatch(addCategory(response.data)); // Dispatch the action to update Redux
            return response.data; // Return added category
        } catch (error) {
            console.error("Error adding category:", error);
            throw error;
        }
    }

    async updateCategory(category: Category) {
        try {
            const response = await axios.put<Category>("http://localhost:8080/admin/category", category);
            store.dispatch(updateCategory(response.data)); // Dispatch the action to update Redux
            return response.data; // Return updated category
        } catch (error) {
            console.error("Error updating category:", error);
            throw error;
        }
    }

    async getAllCategories() {
        try {
            const categories = store.getState().categories.categories;
            if (categories.length === 0) {
                const response = await axios.get<Category[]>("http://localhost:8080/admin/categories");
                store.dispatch(fetchCategories(response.data)); // Dispatch to Redux
                return response.data; // Return the fetched categories
            }
            return categories; // Return the cached categories from Redux store
        } catch (error) {
            console.error("Error fetching categories:", error);
            throw error;
        }
    }

    async getOneCategory(id: number) {
        try {
            let category = store.getState().categories.categories.find(c => c.id === id);
            if (!category) {
                category = (await axios.get<Category>(`http://localhost:8080/admin/category/${id}`)).data;
            }
            return category; // Return found or fetched category
        } catch (error) {
            console.error("Error fetching category:", error);
            throw error;
        }
    }
}

const adminService = new AdminService();
export default adminService;
